nltk/tokenize/casual.py

```diff
@@ -0,0 +1,304 @@
+    |					#   or
+    \([^\s()]*?\([^\s()]+\)[^\s()]*?\) # balanced parens, one level deep: (...(...)...)
+    |
+    \([^\s]+?\)				# balanced parens, non-recursive: (...)
+  )+
+  (?:					# End with:
+    \([^\s()]*?\([^\s()]+\)[^\s()]*?\) # balanced parens, one level deep: (...(...)...)
+    |
+    \([^\s]+?\)				# balanced parens, non-recursive: (...)
+    |					#   or
+    [^\s`!()\[\]{};:'".,<>?«»“”‘’]	# not a space or one of these punct chars
+  )
+  |					# OR, the following to match naked domains:
+  (?:
+  	(?<!@)			        # not preceded by a @, avoid matching foo@_gmail.com_
+    [a-z0-9]+
+    (?:[.\-][a-z0-9]+)*
+    [.]
+    (?:[a-z]{2,13})
+    \b
+    /?
+    (?!@)			        # not succeeded by a @, avoid matching "foo.na" in "foo.na@example.com"
+  )
+"""
+
+# The components of the tokenizer:
+REGEXPS = (
+    URLS,
+    # Phone numbers:
+    r"""
+    (?:
+      (?:            # (international)
+        \+?[01]
+        [\-\s.]*
+      )?
+      (?:            # (area code)
+        [\(]?
+        \d{3}
+        [\-\s.\)]*
+      )?
+      \d{3}          # exchange
+      [\-\s.]*
+      \d{4}          # base
+    )"""
+    ,
+    # ASCII Emoticons
+    EMOTICONS
+    ,
+    # HTML tags:
+    r"""<[^>\s]+>"""
+    ,
+    # ASCII Arrows
+    r"""[\-]+>|<[\-]+"""
+    ,
+    # Twitter username:
+    r"""(?:@[\w_]+)"""
+    ,
+    # Twitter hashtags:
+    r"""(?:\#+[\w_]+[\w\'_\-]*[\w_]+)"""
+    ,
+
+    # Remaining word types:
+    r"""
+    (?:[a-z][a-z'\-_]+[a-z])       # Words with apostrophes or dashes.
+    |
+    (?:[+\-]?\d+[,/.:-]\d+[+\-]?)  # Numbers, including fractions, decimals.
+    |
+    (?:[\w_]+)                     # Words without apostrophes or dashes.
+    |
+    (?:\.(?:\s*\.){1,})            # Ellipsis dots.
+    |
+    (?:\S)                         # Everything else that isn't whitespace.
+    """
+    )
+
+######################################################################
+# This is the core tokenizing regex:
+
+WORD_RE = re.compile(r"""(%s)""" % "|".join(REGEXPS), re.VERBOSE | re.I
+                     | re.UNICODE)
+
+# The emoticon string gets its own regex so that we can preserve case for
+# them as needed:
+EMOTICON_RE = re.compile(EMOTICONS, re.VERBOSE | re.I | re.UNICODE)
+
+# These are for regularizing HTML entities to Unicode:
+HTML_ENTITY_DIGIT_RE = re.compile(r"&#\d+;")
+HTML_ENTITY_ALPHA_RE = re.compile(r"&\w+;")
+AMP = "&amp;"
+
+######################################################################
+
+class TweetTokenizer:
+    """Tokenize Tweets"""
+    def __init__(self, preserve_case=True):
+        self.preserve_case = preserve_case
+
+    def tokenize(self, text):
+        """
+        :param s: str
+        :rtype: list(str)
+        :return: a tokenized list of strings; concatenating this list returns
+        the original string if preserve_case=False
+        """
+        # Try to ensure unicode:
+        try:
+            text = str(text)
```

Comment:
`text = str(text)` is not correct: if unicode text is passed, in Python 2.x it will try to *encode* it to bytes using `sys.getdefaultencoding()`.

I think tokenize function should just require unicode input, and the whole "Try to ensure unicode" block should be removed.